flexmeasures/data/services/utils.py

```diff
@@ -0,0 +1,110 @@
+
+    if callable(
+        getattr(o, "make_hashable", None)
+    ):  # checks if the object o has the method make_hashable
+        return o.make_hashable()
+
+    return o
+
+
+def hash_function_arguments(args, kwags):
+    """Combines the hashes of the args and kargs
+
+    The way to go to do h(x,y) = hash(hash(x) || hash(y)) because it avoid the following:
+
+    1) h(x,y) = hash(x || y), might create a collision if we delete the last n characters of x and we append them in front of y. e.g h("abc", "d") = h("ab", "cd")
+    2) we don't want to sort x and y, because we need the function h(x,y) != h(y,x)
+    3) extra hashing just avoid that we can't decompose the input arguments and track if the same args or kwarg are called several times. More of a security measure I think.
+
+    source: https://crypto.stackexchange.com/questions/55162/best-way-to-hash-two-values-into-one
+    """
+    return make_hash_sha256(
+        make_hash_sha256(args) + make_hash_sha256(kwags)
+    )  # concat two hashes
+
+
+def job_cache(queue):
+    """
+    To avoid recomputing the same task multiple times, this decorator checks if the function was already been called with the
+    same arguments. Input arguments are hashed and stored as Redis keys with the values being the job IDs `input_arguments_hash:job_id`).
+
+    The benefits of using redis to store the input arguments over a local cache, such as LRU Cache, are:
+    1) It will work in distributed environments (in computing clusters), where multiple workers will avoid repeating
+      work as the cache will be shared across them.
+    2) Cached calls are logged, which means that we can easily debug.
+    3) Cache will still be there on restarts.
+    """
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+
+            # checking if force is an input argument of `func`
+            force_new_job_creation = kwargs.pop("force_new_job_creation", False)
+
+            # creating a hash from args and kwargs
+            args_hash = hash_function_arguments(args, kwargs)
+
+            # check if the key hash exists in the redis equeue
+            if (
+                current_app.queues[queue].connection.exists(args_hash)
+                and not force_new_job_creation
+            ):
+                current_app.logger.info(
+                    f"The function {func.__name__} has been called already with the same arguments. Skipping..."
+                )
+
+                # get job id
+                job_id = current_app.queues[queue].connection.get(args_hash).decode()
+
+                job = Job.fetch(
+                    job_id, connection=current_app.queues[queue].connection
+                )  # get job object from job id
+
+                return job  # returning the same job regardless of the status (SUCCESS, FAILED, ...)
+            else:
+                # if the job hasn't been called before or the job has failed -> create job
```

Comment:
No longer a valid comment. I suggest: `# If the job description is new -> create job`